File: carelog/src/utils/dateUtils.ts
import { format, differenceInDays, parseISO } from 'date-fns';

export function today(): string {
  return format(new Date(), 'yyyy-MM-dd');
}

export function formatVisitDate(dateStr: string): string {
  try {
    return format(parseISO(dateStr), 'd MMM yyyy');
  } catch {
    return dateStr;
  }
}

/** Returns the number of calendar days from today until dateStr (negative = past). */
export function getDaysUntil(dateStr: string): number {
  const target = parseISO(dateStr);
  const now = new Date();
  now.setHours(0, 0, 0, 0);
  return differenceInDays(target, now);
}

/** @deprecated Use getDaysUntil */
export const daysUntil = getDaysUntil;

export function isOverdue(dateStr: string): boolean {
  return getDaysUntil(dateStr) < 0;
}

export function formatDaysRemaining(dateStr: string): string {
  const days = getDaysUntil(dateStr);
  if (days < 0) return 'Overdue';
  if (days === 0) return 'Today';
  if (days === 1) return 'Tomorrow';
  return `${days} days left`;
}

/** Returns whole years of age from a YYYY-MM-DD date of birth. Returns 0 if DOB is in the future. */
export function computeAge(dob: string): number {
  const birth = parseISO(dob);
  const now = new Date();
  let age = now.getFullYear() - birth.getFullYear();
  const monthDiff = now.getMonth() - birth.getMonth();
  if (monthDiff < 0 || (monthDiff === 0 && now.getDate() < birth.getDate())) {
    age--;
  }
  return Math.max(0, age);
}

export type ExpiryStatus = 'none' | 'active' | 'expiring' | 'expired';

/**
 * Classifies an insurance/policy expiry date relative to today.
 * `soonDays` controls the "expiring soon" window.
 */
export function getExpiryStatus(
  validUntil: string | undefined | null,
  soonDays = 30,
): { status: ExpiryStatus; label: string } {
  if (!validUntil) return { status: 'none', label: '' };
  const days = getDaysUntil(validUntil);
  if (days < 0) return { status: 'expired', label: 'Expired' };
  if (days === 0) return { status: 'expiring', label: 'Expires today' };
  if (days <= soonDays) return { status: 'expiring', label: `Expires in ${days} day${days === 1 ? '' : 's'}` };
  return { status: 'active', label: `Valid till ${formatVisitDate(validUntil)}` };
}

export function formatCurrency(amount: number | undefined | null, currency = 'INR'): string {
  if (amount == null) return '';
  const symbols: Record<string, string> = { INR: '₹', USD: '$', EUR: '€' };
  const symbol = symbols[currency] ?? currency + ' ';
  return `${symbol}${amount.toLocaleString()}`;
}

File: flutter_version/lib/utils/date_utils.dart
import 'package:intl/intl.dart';

/// Date / currency helpers ported from `../carelog/src/utils/dateUtils.ts`.
/// Dates are ISO strings (`yyyy-MM-dd`) to stay byte-compatible with the RN app.

final DateFormat _ymd = DateFormat('yyyy-MM-dd');
final DateFormat _display = DateFormat('d MMM yyyy');

/// Today as `yyyy-MM-dd`.
String today() => _ymd.format(DateTime.now());

/// Pretty visit/policy date, e.g. `5 Jun 2026`. Falls back to the raw string.
String formatVisitDate(String dateStr) {
  try {
    return _display.format(DateTime.parse(dateStr));
  } catch (_) {
    return dateStr;
  }
}

/// Whole calendar days from today until [dateStr] (negative = in the past).
int getDaysUntil(String dateStr) {
  final target = DateTime.parse(dateStr);
  final targetDay = DateTime(target.year, target.month, target.day);
  final now = DateTime.now();
  final todayDay = DateTime(now.year, now.month, now.day);
  return targetDay.difference(todayDay).inDays;
}

bool isOverdue(String dateStr) => getDaysUntil(dateStr) < 0;

String formatDaysRemaining(String dateStr) {
  final days = getDaysUntil(dateStr);
  if (days < 0) return 'Overdue';
  if (days == 0) return 'Today';
  if (days == 1) return 'Tomorrow';
  return '$days days left';
}

/// Whole years of age from a `yyyy-MM-dd` DOB. 0 if the DOB is in the future.
int computeAge(String dob) {
  final birth = DateTime.parse(dob);
  final now = DateTime.now();
  var age = now.year - birth.year;
  if (now.month < birth.month ||
      (now.month == birth.month && now.day < birth.day)) {
    age--;
  }
  return age < 0 ? 0 : age;
}

/// Expiry classification for an insurance policy (mirrors RN `ExpiryStatus`).
enum ExpiryStatus { none, active, expiring, expired }

/// Result of [getExpiryStatus]: the status plus a ready-to-render label.
class ExpiryInfo {
  final ExpiryStatus status;
  final String label;
  const ExpiryInfo(this.status, this.label);
}

/// Classifies a policy expiry date relative to today. [soonDays] is the
/// "expiring soon" window (default 30, see `insurance.dart`).
ExpiryInfo getExpiryStatus(String? validUntil, {int soonDays = 30}) {
  if (validUntil == null || validUntil.isEmpty) {
    return const ExpiryInfo(ExpiryStatus.none, '');
  }
  final days = getDaysUntil(validUntil);
  if (days < 0) return const ExpiryInfo(ExpiryStatus.expired, 'Expired');
  if (days == 0) return const ExpiryInfo(ExpiryStatus.expiring, 'Expires today');
  if (days <= soonDays) {
    return ExpiryInfo(
        ExpiryStatus.expiring, 'Expires in $days day${days == 1 ? '' : 's'}');
  }
  return ExpiryInfo(ExpiryStatus.active, 'Valid till ${formatVisitDate(validUntil)}');
}

/// `₹1,200` style currency formatting (mirrors RN `formatCurrency`).
String formatCurrency(num? amount, {String currency = 'INR'}) {
  if (amount == null) return '';
  const symbols = {'INR': '₹', 'USD': '\$', 'EUR': '€'};
  final symbol = symbols[currency] ?? '$currency ';
  final formatted = NumberFormat.decimalPattern().format(amount);
  return '$symbol$formatted';
}
